### ISSUE TITLE
LSP stops working after "LSP: Restart Servers"

### ISSUE THREAD

**Issue by deathaxe** [tomv564/LSP#165]
If the "LSP: Restart Servers" command is called from command pallet, the server restarts but nothing works for the open documents. Each request sent to the language-server is responded with an error.

The debug log shows no `didOpen` notification being sent for any document after restart. Even activating/deactivating a view does not fix the issue. All open views must be closed and re-opened.

This issue is related to #145 and definitely an issue of LSP but not the pyls server, which expects `didOpen` to be called before any other request. See last post of #145 for the expected sequence.

```
LSP: starting in C:\Apps\Sublime\Data\Packages
LSP: starting ['pyls']
LSP: request 1: initialize 
LSP: client registered for window 2 {'pyls': <LSP.main.Client object at 0x00000162B38BC518>}
LSP: got json:  {"result":{"capabilities":{"codeActionProvider":true,"codeLensProvider":{"resolveProvider":false},"completionProvider":{"resolveProvider":false,"triggerCharacters":["."]},"documentFormattingProvider": ...
LSP: notify: initialized
LSP: request 2: textDocument/hover 
LSP: got json:  {"error":{"code":-32000,"message":"Server error","data":{"type":"KeyError","args":["file:///C:/Apps/Sublime/Data/Packages/GitGutter/modules/compare.py"],"message":"'file:///C:/Apps/Sublime/Data/Packag ...
LSP: Got error from server:  {'message': 'Server error', 'code': -32000, 'data': {'message': "'file:///C:/Apps/Sublime/Data/Packages/GitGutter/modules/compare.py'", 'args': ['file:///C:/Apps/Sublime/Data/Packages/GitGutter/modules/compare.py'], 'type': 'KeyError'}}
LSP: request 3: textDocument/hover 
LSP: got json:  {"error":{"code":-32000,"message":"Server error","data":{"type":"KeyError","args":["file:///C:/Apps/Sublime/Data/Packages/GitGutter/vcsadapter.py"],"message":"'file:///C:/Apps/Sublime/Data/Packages/Gi ...
LSP: Got error from server:  {'message': 'Server error', 'code': -32000, 'data': {'message': "'file:///C:/Apps/Sublime/Data/Packages/GitGutter/vcsadapter.py'", 'args': ['file:///C:/Apps/Sublime/Data/Packages/GitGutter/vcsadapter.py'], 'type': 'KeyError'}}
```

**Comment by tomv564** [comment]
We can definitely improve this behaviour by sending a `start_active_view` after restarting. Restarting should not be a regular thing, so I hope any pyls issues causing you to restart are reported?

**Comment by deathaxe** [comment]
The starting point of it was, I had several python files open in a multi column layout. All working well, when I opened LSP/main.py to investigate on the issues with the completions stuff. 

After saving main.py the server is closed and restarted. The main.py recovered, everything working. The other views caused server error when hovering a keyword. So I remembered issue #145 as we were not finally able to find the reason/responsibility there. Saw a chance here to do so. Ok, clicking inside a view triggered didOpen by the on_activate event. Works fine, than.

I rememberd the RestartServerCommand, so made sure pyls works for all views and called the command from command pallet. From this point on I was no longer able to make things work by activating/deactivating the views. pyls never recovered, always responded with error.

Same situation while working on the shutdown stuff. I tried to simulate several situations in which a server might close itself by a crash or exits on its own rather than by "exit" notification. Just to see whether LSP is able to handle different (error) situations well. As soon as a server exits on its own or is closed via TaskManager LSP is no longer able to recover reliably. I don't want to blame someone for responsibility. There are quite a bunch of situations, I wouldn't have expected at all.

Ok, one could consider all those situations as uncommon edge cases. But in the end LSP should be able to handle server crashes gracefully and recover correctly after its restart. Otherwise a user would always need to restart ST to do so. At least the "RestartServerCommand" should recover everything as it can be called by each user in any situation.

I honestly wouldn't have expected so much things/situations/edge cases/... to be considered, when working with such kind of external services.
